sources/api/netdog/src/cli/write_resolv_conf.rs

```diff
@@ -1,28 +1,177 @@
+            }),
+            dhcp6: Some(Dhcp6Section {
+                use_dns,
+                use_domains,
+            }),
+        })
+    }
+}
+
+// Gather the DNS settings if they exist, and if so create interface level drop-ins to turn off
+// DNS-via-DHCP for either name servers or search domains.  Then write a systemd-resolved
+// configuration drop-in with the global DNS settings.
+//
+// For example, if we have DNS search domain settings from the API, we don't need to get them from
+// DHCP, so the interface drop-in will turn them off and systemd-resolved will use the global
+// setting from the API which will be set in the systemd-resolved configuration file.
+#[cfg(net_backend = "systemd-networkd")]
+fn handle_dns_settings(_primary_interface: String) -> Result<()> {
+    // The DNS settings config file is always created since it is a configuration file in the API.
+    // If it is empty, don't create any interface drop-ins
+    let dns_settings = DnsSettings::from_config().context(error::GetDnsSettingsSnafu)?;
+    let interface_drop_in = match InterfaceDNSDropIn::from_dns_settings(&dns_settings) {
+        Some(d) => d,
+        None => {
+            eprintln!("No DNS configuration to write");
+            return Ok(());
+        }
+    };
+
+    // For each configured interface, create the drop-in directory and file It is possible for net
+    // config to be empty/None; if that is the case we need to fail here since we won't know which
+    // interfaces for which to write drop-ins.
+    let (maybe_net_config, _) = fetch_net_config()?;
+    let net_config = maybe_net_config.context(error::MissingNetworkConfigSnafu)?;
+    for interface in net_config.interfaces() {
+        // Remove the colons since the ID might be a MAC address
+        let name = interface.to_string().replace(':', "");
+        let dropin_dirname = format!("10-{}.network.d", name);
+        let dropin_dir_path = Path::new(NETWORKD_CONFIG_DIR).join(&dropin_dirname);
+        fs::create_dir_all(&dropin_dir_path).context(error::DropInDirCreateSnafu {
+            path: &dropin_dir_path,
+        })?;
+
+        let mut dropin_file_path = dropin_dir_path.join("10-dns");
+        dropin_file_path.set_extension("conf");
+        fs::write(&dropin_file_path, &interface_drop_in.to_string()).context(
+            error::DropInFileWriteSnafu {
+                path: dropin_file_path,
+            },
+        )?;
+    }
+
+    // Write the systemd-resolved drop-in which will contain the DNS settings from the API
+    dns_settings
+        .write_resolved_dropin()
+        .context(error::ResolvConfWriteFailedSnafu)?;
+
+    //** Note: order of operations is important here!
+    // After all the above file writes have completed successfully, restart systemd-networkd and
+    // then systemd-resolved.  Restarting networkd first ensures that desired interface-level DNS
+    // settings are set before we restart systemd-resolved to pick up the global settings.
```

Comment:
Is the order really important? It seems like resolved and networkd should end up in agreement either way, and if they don't then there might be a race condition.

Just restarting the services one after another doesn't necessarily mean they are ready and will reach a consistent state; for example, if networkd is sending per-interface updates to resolved when it starts and then resolved goes away during that procedure, then networkd may not retry or continue for other interfaces.